### PR TITLE
Update module to use authentification V2 for function apps

### PR DESCRIPTION
**Pull request by ArtiomMatiom**
azapi_update_resource -> was causing troubles when provider version were changing. 

Updated the azurerm provider version which added the functionality for Managed Identity authentification in a linux function app

### CODE REVIEW

File: modules/azure/function_app_linux_managed_identity/main.tf [hunk 1]
```diff
@@ -4,7 +4,7 @@ terraform {
   required_providers {
     azurerm = {
       source  = "hashicorp/azurerm"
-      version = "~> 3.40"
+      version = "=3.47.0"
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/main.tf:7]:
```suggestion
      version = "~> 3.47"
```

Don't use a fixed version.

File: modules/azure/function_app_linux_managed_identity/variables.tf
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/variables.tf:1]: Fix the linting https://github.com/recognizegroup/terraform/actions/runs/4428578390/jobs/7767863732

File: modules/azure/function_app_linux_managed_identity/main.tf [hunk 2]
```diff
@@ -176,11 +160,12 @@ resource "azurerm_monitor_diagnostic_setting" "diagnostic_setting" {
   target_resource_id         = azurerm_linux_function_app.function_app.id
   log_analytics_workspace_id = var.log_analytics_workspace_id
 
-  dynamic "enabled_log" {
-    for_each = data.azurerm_monitor_diagnostic_categories.diagnostic_categories[0].log_category_types
+  dynamic "log" {
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/main.tf:163]: Is there any reason why this and log_category_types are reverted back to the deprecated attributes?

File: modules/azure/function_app_linux_managed_identity/variables.tf [hunk 3]
```diff
@@ -123,3 +123,17 @@ variable "ip_restriction" {
   description = "A List of objects representing ip restrictions."
   default     = null
 }
+
+variable "authetification_settings" {
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/variables.tf:127]:
```suggestion
variable "authentication_settings" {
```

File: modules/azure/function_app_linux_managed_identity/main.tf [hunk 4]
```diff
@@ -61,50 +65,30 @@ resource "azurerm_linux_function_app" "function_app" {
     }
   }
 
+  auth_settings_v2 {
+    auth_enabled           = true
+    require_authentication = var.authetification_settings.require_authentication == null ? false : var.authetification_settings.require_authentication
+    unauthenticated_action = var.authetification_settings.unauthenticated_action == null ? null : var.authetification_settings.unauthenticated_action
+    excluded_paths         = var.authetification_settings.excluded_paths == null ? [] : var.authetification_settings.excluded_paths
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/main.tf:72]:
```suggestion
    require_authentication = var.authentication_settings.require_authentication == null ? false : var.authentication_settings.require_authentication
    unauthenticated_action = var.authentication_settings.unauthenticated_action == null ? null : var.authentication_settings.unauthenticated_action
    excluded_paths         = var.authentication_settings.excluded_paths == null ? [] : var.authentication_settings.excluded_paths
```

**Review by tom-reinders — CHANGES_REQUESTED** [review]

File: modules/azure/function_app_linux_managed_identity/variables.tf
Comment (ArtiomMatiom) [reply modules/azure/function_app_linux_managed_identity/variables.tf:1]: Done

File: modules/azure/function_app_linux_managed_identity/main.tf [hunk 5]
```diff
@@ -181,6 +165,7 @@ resource "azurerm_monitor_diagnostic_setting" "diagnostic_setting" {
 
     content {
       category = enabled_log.value
+      enabled  = true
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/main.tf:168]: You still forgot to revert this change, this is part of the change to enabled_log

File: modules/azure/function_app_linux_managed_identity/main.tf [hunk 6]
```diff
@@ -61,50 +65,30 @@ resource "azurerm_linux_function_app" "function_app" {
...
+
+    active_directory_v2 {
+      client_id                  = local.should_create_app ? azuread_application.application[0].application_id : var.managed_identity_provider.existing.client_id
+      client_secret_setting_name = "MICROSOFT_PROVIDER_AUTHENTICATION_SECRET"
+      tenant_auth_endpoint       = "https://login.microsoftonline.com/v2.0/${data.azurerm_client_config.current.tenant_id}/"
+      allowed_audiences          = local.allowed_audiences
+    }
+
+    login {
+      // Bug within terraform module it just requires it
```
Comment (tom-reinders) [inline modules/azure/function_app_linux_managed_identity/main.tf:82]: By any chance is there an ticket of this bug in for example https://github.com/hashicorp/terraform-provider-azurerm?
If so link the ticket in the comment so we can easily see if the bug has been fixed.

Comment (ArtiomMatiom) [reply modules/azure/function_app_linux_managed_identity/main.tf:82]: Added commented link to a issue I created

**Review by tom-reinders — APPROVED** [review]

**Comment by tom-reinders** [comment]
Adding the issue link here so the 2 are linked to each other hashicorp/terraform-provider-azurerm#21002

**Review by tom-reinders — APPROVED** [review]
